include/dao/db_manager.js

```diff
@@ -185,6 +199,91 @@ module.exports = function DBManagerModule(pb) {
                 };
                 cb(err, result);
             });
+        }
+
+
+        /**
+         * Sorts through all created indices and drops any index not declared in pb.config
+         * @method dropUnconfiguredIndices
+         * @param {Array} procedures
+         * @param {Function} cb
+         */
+        this.dropUnconfiguredIndices = function(procedures, cb) {
+            var self = this;
+            //to prevent a cirular dependency we do the require for DAO here.
+            var DAO = require('./dao.js')(pb);
+            this.getStoredIndices(function(err, storedIndices) {
+                var dao = new DAO();
+
+                var tasks = util.getTasks(storedIndices, function(indices, i) {
+                    return function(callback) {
+                        var index = indices[i];
+                        var filteredIndex = procedures.filter(function(procedure) {
+                            var ns = pb.config.db.name + '.' + procedure.collection;
+                            var result = ns === index.ns && self.compareIndices(index, procedure);
+                            return result;
+                        });
+                        var indexCollection = index.ns.split('.')[1];
+                        if(index.name !== '_id_' && indexCollection !== 'session' && (filteredIndex.length === 0 || util.isNullOrUndefined(filteredIndex))) {
+                            dao.dropIndex(indexCollection, index.name, function(err, result) {
+                                if(util.isError(err)) {
+                                    pb.log.error('DBManager: Failed to drop undeclared INDEX=[%s] RESULT=[%s] ERROR[%s]', JSON.stringify(index), util.inspect(result), err.stack);
+                                }
+                                callback(err, result);
+                            });
+                        }
+                        else {
+                            callback();
+                        }
```

Comment:
This is done as well with a comment explaining what is being evaluated.

Ben Tidwell
Software Engineer III | Talent Network
CareerBuilder
5550-A Peachtree Pkwy, Norcross GA 30092

From: Brian <notifications@github.com<mailto:notifications@github.com>>
Reply-To: pencilblue/pencilblue <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Sunday, July 12, 2015 at 7:10 PM
To: pencilblue/pencilblue <pencilblue@noreply.github.com<mailto:pencilblue@noreply.github.com>>
Cc: Ben Tidwell <ben.tidwell@careerbuilder.com<mailto:ben.tidwell@careerbuilder.com>>
Subject: Re: [pencilblue] Multi Tenancy (#680)


In include/dao/db_manager.js<https://github.com/pencilblue/pencilblue/pull/680#discussion_r34426850>:

> +                            var ns = pb.config.db.name + '.' + procedure.collection;
> +                            var result = ns === index.ns && self.compareIndices(index, procedure);
> +                            return result;
> +                        });
> +                        var indexCollection = index.ns.split('.')[1];
> +                        if(index.name !== '_id_' && indexCollection !== 'session' && (filteredIndex.length === 0 || util.isNullOrUndefined(filteredIndex))) {
> +                            dao.dropIndex(indexCollection, index.name, function(err, result) {
> +                                if(util.isError(err)) {
> +                                    pb.log.error('DBManager: Failed to drop undeclared INDEX=[%s] RESULT=[%s] ERROR[%s]', JSON.stringify(index), util.inspect(result), err.stack);
> +                                }
> +                                callback(err, result);
> +                            });
> +                        }
> +                        else {
> +                            callback();
> +                        }


Use the inverse of the if statement. It will prevent the need to use the "else" statement and reduce nesting depth.

if(index.name === '_id_' || indexCollection === 'session' || (filteredIndex.length !== 0 && !util.isNullOrUndefined(filteredIndex))) {
  return callback();
}

//proceed to drop index here


-
Reply to this email directly or view it on GitHub<https://github.com/pencilblue/pencilblue/pull/680/files#r34426850>.